stop using requirementslib models

matteius opened the pull request:
The requirementslib models are a nested mess.  I fixed the issue of wheel extras not being installed and broke a bunch of other stuff in the process the other night, but it dawned on me that there is a lot of translating between pip lines and data models and  back and forth and it really is messy and doesn't make much sense.

This is the start of re-establishing what interfaces are actually needed to get between Pipfile, Pipfile.lock and pip lines.   The preferred data model to use is pip's InstallRequirement since requirementslib was using it under the hood anwyay as `ireq`.

There are still a number of edge cases to work out, but I am opening  this draft PR now because I want to see if I actually broke the test runner or just my local environment.

At a high level this is what I am thinking but it requires dropping `--skip-lock` the rest of the way first:
![image](https://github.com/pypa/pipenv/assets/479892/9d2b1e92-fbf7-491c-85f3-b9f26636e646)

Fixes #5817
Fixes #5789
Fixes #5696
Fixes #5626
Fixes #5595 
Fixes #5490
Fixes #5716 (probably)

### The checklist

* [X] Associated issue
* [X] A news fragment in the `news/` directory to describe this fix with the extension `.bugfix.rst`, `.feature.rst`, `.behavior.rst`, `.doc.rst`. `.vendor.rst`. or `.trivial.rst` (this will appear in the release changelog). Use semantic line breaks and name the file after the issue number or the PR #.

oz123 commented on pipenv/vendor/requirementslib/models/requirements.py:
```diff
@@ -2308,7 +2298,7 @@ def get_hashes_as_pip(self, as_list: bool = False) -> Union[str, List[str]]:
         return self.get_hashes_as_pip()
 
-    @cached_property
+    @property
```
why remove cached_property here?

DanielSwain commented:
@matteius To address [your request to me](https://github.com/pypa/pipenv/issues/5798#issuecomment-1659572658) to test this code (and also your request [to @joaomcarlos](https://github.com/pypa/pipenv/issues/5791)), I cloned the full pipenv repo and then installed your ```draft-no-reqlib``` branch on my local machine with the following (after uninstalling the previously installed ```pipenv```):
```pip3 install git+https://github.com/DanielSwain/pipenv.git@draft-no-reqlib```
I ran ```pipenv update``` in the project that has a private repo as part of its dependencies.  It failed in attempting the lock.  Below are the full results from the point of running ```python3 -m pipenv update```.  I hope this is helpful.

```
Running $ pipenv lock then $ pipenv sync.
Locking [packages] dependencies...
Building requirements...
Resolving dependencies...
Resolving dependencies...
✘ Locking Failed!
⠴ Locking...
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/pipenv/patched/pip/_vendor/packaging/requirements.py", line 102, in __init__
    req = REQUIREMENT.parseString(requirement_string)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/patched/pip/_vendor/pyparsing/util.py", line 256, in _inner
    return fn(self, *args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/patched/pip/_vendor/pyparsing/core.py", line 1190, in parse_string
    raise exc.with_traceback(None)
pipenv.patched.pip._vendor.pyparsing.exceptions.ParseException: Expected string_end, found ':'  (at char 3), (line:1, col:4)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/pipenv/patched/pip/_internal/req/constructors.py", line 341, in _parse_req_string
    req = get_requirement(req_as_string)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/patched/pip/_internal/utils/packaging.py", line 45, in get_requirement
    return Requirement(req_string)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/patched/pip/_vendor/packaging/requirements.py", line 104, in __init__
    raise InvalidRequirement(
pipenv.patched.pip._vendor.packaging.requirements.InvalidRequirement: Parse error at "'://git@b'": Expected string_end
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/pipenv/resolver.py", line 699, in <module>
    main()
  File "/usr/local/lib/python3.8/dist-packages/pipenv/resolver.py", line 685, in main
    _main(
  File "/usr/local/lib/python3.8/dist-packages/pipenv/resolver.py", line 665, in _main
    resolve_packages(
  File "/usr/local/lib/python3.8/dist-packages/pipenv/resolver.py", line 642, in resolve_packages
    results = clean_results(results, resolver, project, category)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/resolver.py", line 577, in clean_results
    entry_dict = translate_markers(entry.get_cleaned_dict())
  File "/usr/local/lib/python3.8/dist-packages/pipenv/resolver.py", line 246, in get_cleaned_dict
    if self.entry.extras != self.lockfile_entry.extras:
  File "/usr/local/lib/python3.8/dist-packages/pipenv/resolver.py", line 549, in __getattribute__
    return super().__getattribute__(key)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/resolver.py", line 268, in lockfile_entry
    self._lockfile_entry = self.make_requirement(self.name, self.lockfile_dict)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/resolver.py", line 146, in make_requirement
    return from_pipfile(name, entry)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/utils/dependencies.py", line 901, in from_pipfile
    install_req, markers = install_req_from_pipfile(name, pipfile)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/utils/dependencies.py", line 887, in install_req_from_pipfile
    install_req = expansive_install_req_from_line(
  File "/usr/local/lib/python3.8/dist-packages/pipenv/utils/dependencies.py", line 842, in expansive_install_req_from_line
    parts = parse_req_from_line(name, line_source)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/patched/pip/_internal/req/constructors.py", line 369, in parse_req_from_line
    req: Optional[Requirement] = _parse_req_string(req_as_string)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/patched/pip/_internal/req/constructors.py", line 355, in _parse_req_string
    raise InstallationError(msg)
pipenv.patched.pip._internal.exceptions.InstallationError: Invalid requirement: 'ssh://git@bitbucket.org/OUR_ORGANIZATION/OUR_PRIVATE_REPO.git'
Hint: It looks like a path. File 'ssh://git@bitbucket.org/OUR_ORGANIZATION/OUR_PRIVATE_REPO.git' does not exist.

Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/__main__.py", line 4, in <module>
    cli()
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/cli/options.py", line 57, in main
    return super().main(*args, **kwargs, windows_expand_args=False)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/decorators.py", line 84, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/cli/command.py", line 570, in update
    do_update(
  File "/usr/local/lib/python3.8/dist-packages/pipenv/routines/update.py", line 57, in do_update
    do_lock(
  File "/usr/local/lib/python3.8/dist-packages/pipenv/routines/lock.py", line 65, in do_lock
    venv_resolve_deps(
  File "/usr/local/lib/python3.8/dist-packages/pipenv/utils/resolver.py", line 781, in venv_resolve_deps
    c = resolve(cmd, st, project=project)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/utils/resolver.py", line 654, in resolve
    raise RuntimeError("Failed to lock Pipfile.lock!")
RuntimeError: Failed to lock Pipfile.lock!

root@MY-DEV-MACHINE:/project_directory# pip3 uninstall pipenv
Found existing installation: pipenv 2023.7.24.dev0
Uninstalling pipenv-2023.7.24.dev0:
  Would remove:
    /usr/local/bin/pipenv
    /usr/local/bin/pipenv-resolver
    /usr/local/lib/python3.8/dist-packages/pipenv-2023.7.24.dev0.dist-info/*
    /usr/local/lib/python3.8/dist-packages/pipenv/*
Proceed (Y/n)? Y
  Successfully uninstalled pipenv-2023.7.24.dev0

root@MY-DEV-MACHINE:/var/sites/paulownia# pip3 install pipenv
Collecting pipenv
  Obtaining dependency information for pipenv from https://files.pythonhosted.org/packages/91/a7/97ecc0da97fff327eef7b47cfefc8a5e1f5b20cf45cd1b119c2392199647/pipenv-2023.7.23-py3-none-any.whl.metadata
  Downloading pipenv-2023.7.23-py3-none-any.whl.metadata (14 kB)
Requirement already satisfied: certifi in /usr/lib/python3/dist-packages (from pipenv) (2019.11.28)
Requirement already satisfied: setuptools>=67.0.0 in /usr/local/lib/python3.8/dist-packages (from pipenv) (67.7.2)
Requirement already satisfied: virtualenv-clone>=0.2.5 in /usr/local/lib/python3.8/dist-packages (from pipenv) (0.5.4)
Requirement already satisfied: virtualenv>=20.17.1 in /usr/local/lib/python3.8/dist-packages (from pipenv) (20.20.0)
Requirement already satisfied: distlib<1,>=0.3.6 in /usr/local/lib/python3.8/dist-packages (from virtualenv>=20.17.1->pipenv) (0.3.6)
Requirement already satisfied: filelock<4,>=3.4.1 in /usr/local/lib/python3.8/dist-packages (from virtualenv>=20.17.1->pipenv) (3.9.0)
Requirement already satisfied: platformdirs<4,>=2.4 in /usr/local/lib/python3.8/dist-packages (from virtualenv>=20.17.1->pipenv) (3.0.0)
Downloading pipenv-2023.7.23-py3-none-any.whl (2.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.8/2.8 MB 9.5 MB/s eta 0:00:00
DEPRECATION: distro-info 0.23ubuntu1 has a non-standard version number. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of distro-info or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063
Installing collected packages: pipenv
Successfully installed pipenv-2023.7.23

root@MY-DEV-MACHINE:/project_directory# python3 -m pipenv update
Loading .env environment variables...
Running $ pipenv lock then $ pipenv sync.
Locking [packages] dependencies...
Building requirements...
Resolving dependencies...
Resolving dependencies...
✔ Success!
Locking [dev-packages] dependencies...
Updated Pipfile.lock (33ef1338______________________________________________)!
Installing dependencies from Pipfile.lock (8fa064)...
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.
All dependencies are now up-to-date!
```

matteius commented:
@DanielSwain that is helpful to know there is an edge case but I likely need more information to accurately access.   My guess is you have a Pipfile entry for a git requirement and the key starts with ssh://?   If so could you try appending `git+ssh://` to that Pipfile entry and try to relock?  This would be helpful information to know, perhaps some VCS Pipfile entries work on the old version but would be generated slightly differently on the new version.

matteius commented:
@DanielSwain thanks again for checking this -- I just pushed a change that should cover this older style vcs Pipfile entry that is missing the {vcs}+ specifier.  Let me know if you get a chance to re-check it.

oz123 commented on pipenv/utils/safe_ast.py:
```diff
@@ -0,0 +1,148 @@
+import ast
```
Can you please add a short comment describing the purpose of the new module?

matteius replied:
Its not really that new, its just an improvement/simplification over what requirementslib did before in terms of ast parsing:  https://github.com/sarugaku/requirementslib/blob/main/src/requirementslib/models/setup_info.py#L442-L777

If you recall:  https://github.com/pypa/pipenv/pull/5357

Basically setup.py has unsafe code that should not be evaluated during resolution (hence why the community has been moving towards pyproject.toml and setup.cfg).

oz123 commented on pipenv/utils/resolver.py:
```diff
@@ -550,24 +300,20 @@ def default_constraint_file(self):
 
     @property
     def pip_options(self):
```
The past usage of :
```
if self._pip_options is None:

    ....
    
```
Suggest that this was called multiple times. Are we calling this multiple times still?
If so, could this be a cached property?

matteius replied:
It needs to be re-evaluated fresh each time.

matteius replied:
The problem with some of the past cached properties is they were leading to the need to have other bad code such as:  https://github.com/pypa/pipenv/pull/5793/files#diff-62e85156fc3b274046cc03dabfe6e3ceead503f330b98fb3611645216b4b6e1fL685

matteius replied:
![image](https://github.com/pypa/pipenv/assets/479892/4d13b0dc-8dbe-4be4-8949-7e3a562fb031)

DanielSwain commented:
@matteius I installed your latest from your `draft-no-reqlib branch`.  I `python3 -m pipenv update` both before and after adding `git+` in front of `ssh` in the Pipfile, and I got the same error both times:

```
root@MY_DEV_MACHINE:/my/project# python3 -m pipenv update
Loading .env environment variables...
Running $ pipenv lock then $ pipenv sync.
Locking [packages] dependencies...
Building requirements...
Resolving dependencies...
✘ Locking Failed!
⠦ Locking...
ERROR:pip.subprocessor:[present-rich] git fetch -q --tags exited with 128
[ResolutionFailure]:   File "/usr/local/lib/python3.8/dist-packages/pipenv/resolver.py", line 665, in _main
[ResolutionFailure]:       resolve_packages(
[ResolutionFailure]:   File "/usr/local/lib/python3.8/dist-packages/pipenv/resolver.py", line 632, in resolve_packages
[ResolutionFailure]:       results, resolver = resolve(
[ResolutionFailure]:   File "/usr/local/lib/python3.8/dist-packages/pipenv/resolver.py", line 612, in resolve
[ResolutionFailure]:       return resolve_deps(
[ResolutionFailure]:   File "/usr/local/lib/python3.8/dist-packages/pipenv/utils/resolver.py", line 842, in resolve_deps
[ResolutionFailure]:       results, hashes, internal_resolver = actually_resolve_deps(
[ResolutionFailure]:   File "/usr/local/lib/python3.8/dist-packages/pipenv/utils/resolver.py", line 617, in actually_resolve_deps
[ResolutionFailure]:       resolver.resolve()
[ResolutionFailure]:   File "/usr/local/lib/python3.8/dist-packages/pipenv/utils/resolver.py", line 441, in resolve
[ResolutionFailure]:       raise ResolutionFailure(message=str(e))
[pipenv.exceptions.ResolutionFailure]: Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.
  You can use $ pipenv run pip install <requirement_name> to bypass this mechanism, then run $ pipenv graph to inspect the versions actually installed in the virtualenv.
  Hint: try $ pipenv lock --pre if it is a pre-release dependency.
ERROR: git fetch -q --tags exited with 128

Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/__main__.py", line 4, in <module>
    cli()
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/cli/options.py", line 57, in main
    return super().main(*args, **kwargs, windows_expand_args=False)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/decorators.py", line 84, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/vendor/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/cli/command.py", line 570, in update
    do_update(
  File "/usr/local/lib/python3.8/dist-packages/pipenv/routines/update.py", line 57, in do_update
    do_lock(
  File "/usr/local/lib/python3.8/dist-packages/pipenv/routines/lock.py", line 65, in do_lock
    venv_resolve_deps(
  File "/usr/local/lib/python3.8/dist-packages/pipenv/utils/resolver.py", line 783, in venv_resolve_deps
    c = resolve(cmd, st, project=project)
  File "/usr/local/lib/python3.8/dist-packages/pipenv/utils/resolver.py", line 654, in resolve
    raise RuntimeError("Failed to lock Pipfile.lock!")
RuntimeError: Failed to lock Pipfile.lock!
```

matteius commented:
@DanielSwain Can you tell me if your vcs Pipfile entry has any keys besides the git key?   Specifically a ref or branch name?   If it does, I can probably come up with a fix, but if that bit of information is missing, I think it will end up being a backwards incompatibility or somehow it needs to determine the default branch when none is given.   So for example, I found that I had to add `@branch_name_or_ref` at the end of the vcs entry or install line to get it to work.

```
matte@LAPTOP-N5VSGIBD MINGW64 ~/Projects/pipenv-triage/vcs_reqs
$ pipenv install git+ssh://git@github.com/pypa/pipenv.git@main
Installing git+ssh://git@github.com/pypa/pipenv.git@main...
Resolving git+ssh://git@github.com/pypa/pipenv.git@main...
Added pipenv to Pipfile's [packages] ...
Installation Succeeded
Pipfile.lock (5c9a38) out of date, updating to (379467)...
Locking [packages] dependencies...
Building requirements...
Resolving dependencies...
Success!
Locking [dev-packages] dependencies...
Updated Pipfile.lock (aefcf91d1e13a8dc19a9835275b11bf6712875db3ad4ed48bca85b2fb3379467)!
Installing dependencies from Pipfile.lock (379467)...
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.

matte@LAPTOP-N5VSGIBD MINGW64 ~/Projects/pipenv-triage/vcs_reqs
$ cat Pipfile
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
pipenv = {git = "git+ssh://git@github.com/pypa/pipenv.git@main"}

[dev-packages]

[requires]
python_version = "3.9"

matte@LAPTOP-N5VSGIBD MINGW64 ~/Projects/pipenv-triage/vcs_reqs
$ pipenv update
Running $ pipenv lock then $ pipenv sync.
Locking [packages] dependencies...
Building requirements...
Resolving dependencies...
Success!
Locking [dev-packages] dependencies...
Updated Pipfile.lock (aefcf91d1e13a8dc19a9835275b11bf6712875db3ad4ed48bca85b2fb3379467)!
Installing dependencies from Pipfile.lock (379467)...
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.
All dependencies are now up-to-date!
```

DanielSwain commented:
@matteius, here is the format of the private library entry in the Pipfile:
`our-private-library = {ref = "master", git = "git+ssh://git@bitbucket.org/our_organization/our-private-library.git"}`

DanielSwain commented:
@matteius When you get hashing on VCS repos re-implemented, just let me know.  I'd be glad to test that for you.

matteius commented:
@DanielSwain I got the VCS lockfile hashing sorted out now.  👍

kalebmckale commented:
Hey @matteius I'm wondering if there is a way to get more details/output when a dependency resolution check fails. I'm having an issue getting Pipfile with particular private PyPI repo packages to lock, and even with `--verbose` I cannot figure out why this is the case. If I `pip install` the packages, there are no dependency conflicts. If I create a requirements file using `pip freeze` after manually installing everything I can then create a lock file from the requirements file. However, the standard unpinned Pipfile will not lock. Any hidden options or ideas would be greatly helpful.

matteius commented:
@kalebmckale I just pushed a commit to this branch which gets the `-v` or `--verbose` flag working with the right level of pip output from the resolver now.  Please give it a try.

kalebmckale commented:
@matteius I tested with the version. The verbose output was very helpful. I think I know what the problem is, but I'm not sure why it's happening. Is this the thread to discuss it or should I create a new issue since it seems to be behavior independent of the current changes?

matteius commented:
@kalebmckale please open a new issue for that then.

kalebmckale commented:
> @kalebmckale please open a new issue for that then.

Done!
